File: C# - Typing/Implementation.cs
public interface ICounter {
    int       Initial { get; }
    int       Value   { get; }
    ICounter  Reset();
    ICounter  Increment();
    ICounter  Decrement();
    ICounter  Add(int by);
    ICounter  Sub(int by);
    ICounter  Set(int by);
    int       Get();
}


public class Counter: ICounter {
    
    public int Initial { get; private set; }
    public int Value   { get; private set; }

    public Counter(int inival) { 
        Initial = inival; 
        Value   = inival; 
    }

    public ICounter  Reset()     { Value  = Initial;  return this;  }
    public ICounter  Increment() { Value ++;          return this;  }
    public ICounter  Decrement() { Value --;          return this;  }
    public ICounter  Add(int by) { Value += by;       return this;  }
    public ICounter  Sub(int by) { Value -= by;       return this;  }
    public ICounter  Set(int by) { Value  = by;       return this;  }
    public int       Get()       {                    return Value; }
}

File: C++ - Typing/Implementation.cpp
class ICounter {
    public:
        virtual ICounter& reset()     = 0;
        virtual ICounter& increment() = 0;
        virtual ICounter& decrement() = 0;
        virtual ICounter& add(int by) = 0;
        virtual ICounter& sub(int by) = 0;
        virtual ICounter& set(int by) = 0;
        virtual int get()       const = 0;
        virtual ~ICounter()           = default;  // Virtual destructor for cleanup
};


class Counter: public ICounter {
    private:
        int initial;  
        int value;

    public:
        Counter(int iniVal): 
            initial(iniVal), 
              value(iniVal) {}

        ICounter& reset()       override { value  = initial;  return *this;  }
        ICounter& increment()   override { value ++;          return *this;  }
        ICounter& decrement()   override { value --;          return *this;  }
        ICounter& add(int by)   override { value += by;       return *this;  }
        ICounter& sub(int by)   override { value -= by;       return *this;  }
        ICounter& set(int by)   override { value  = by;       return *this;  }
        int       get() const   override {                    return  value; }
};